### ISSUE TITLE
external-check missing server ip using resolvers (haproxy 2.1.4)

### ISSUE THREAD

**Issue by dfex55** [haproxy/haproxy#577]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.1.4 2020/04/02 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.4.html
Usage : haproxy [-f <cfgfile|cfgdir>]* [ -vdVD ] [ -n <maxconn> ] [ -N <maxpconn> ]
        [ -p <pidfile> ] [ -m <max megs> ] [ -C <dir> ] [-- <cfgfile>*]
        -v displays version ; -vv shows known build options.
        -d enters debug mode ; -db only disables background mode.
        -dM[<byte>] poisons memory with <byte> (defaults to 0x50)
        -V enters verbose mode (disables quiet mode)
        -D goes daemon ; -C changes to <dir> before loading files.
        -W master-worker mode.
        -q quiet mode : don't display messages
        -c check mode : only check config files and exit
        -n sets the maximum total # of connections (uses ulimit -n)
        -m limits the usable amount of memory (in MB)
        -N sets the default, per-proxy maximum # of connections (0)
        -L set local peer name (default to hostname)
        -p writes pids of all children to this file
        -de disables epoll() usage even when available
        -dp disables poll() usage even when available
        -dS disables splice usage (broken on old kernels)
        -dG disables getaddrinfo() usage
        -dR disables SO_REUSEPORT usage
        -dr ignores server address resolution failures
        -dV disables SSL verify on servers side
        -sf/-st [pid ]* finishes/terminates old pids.
        -x <unix_socket> get listening sockets from a unix socket
        -S <bind>[,<bind options>...] new master CLI


Linux baf9ca797d7e 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 GNU/Linux
```

## What's the configuration?

```
global
    external-check


resolvers docker
    nameserver dns 127.0.0.11:53

defaults
    log global
    mode tcp
    retries 2
    timeout client 30m
    timeout connect 4s
    timeout server 30m
    timeout check 5s

listen stats
    mode http
    bind *:7000
    maxconn 10
    stats enable
    stats uri /
    stats refresh 10s

frontend service_frontend
    mode tcp
    bind *:1111
    maxconn 1000
    default_backend service_backend

backend service_backend
    mode tcp
    fullconn 1000
    option external-check
    external-check command /srv/customcheck
    default-server init-addr libc,last,none
    server instance-1 my-named-instance-1:1111 maxconn 1000 check inter 1s resolvers docker
    server instance-2 my-named-instance-2:1111 maxconn 1000 check inter 1s resolvers docker
```

## The external-check script?
```
#!/bin/bash

echo params: $@
echo HAPROXY_SERVER_ADDR=$HAPROXY_SERVER_ADDR
echo HAPROXY_SERVER_PORT=$HAPROXY_SERVER_PORT
echo ========================

if [ "$HAPROXY_SERVER_ADDR" == "" ]; then
  exit 1
fi

.....
```

## Steps to reproduce the behavior

1. start haproxy before any other service-instance is up
2. start a service instance
3. watch the log output of the external-check script

## Actual behavior

After starting haproxy, all backend upstreams are marked down. When a service instance comes up, haproxy will resolv its ip and starts calling the external-check script.

external-check script will always fail, because ip and port is missing (in arguments and environment). see output of check script / log of haproxy:

```
[WARNING] 097/125714 (1) : parsing [/usr/local/etc/haproxy/haproxy.cfg] : 'server instance-1' : could not resolve address my-named-instance-1', disabling server.

[WARNING] 097/125714 (1) : parsing [/usr/local/etc/haproxy/haproxy.cfg] : 'server instance-2' : could not resolve address my-named-instance-2', disabling server.

[NOTICE] 097/125714 (1) : New worker #1 (6) forked

[WARNING] 097/125744 (6) : service_backend/instance-1 changed its IP from  to 10.44.0.3 by docker/dns.

[WARNING] 097/125744 (6) : Server service_backend/instance-1 ('my-named-instance-1') is UP/READY (resolves again).

[WARNING] 097/125744 (6) : Server service_backend/instance-1 administratively READY thanks to valid DNS answer.

params: NOT_USED NOT_USED

HAPROXY_SERVER_ADDR=

HAPROXY_SERVER_PORT=

========================
```
As you can see, dns lookup worked (ip is valid), but ip is not used for the script

## Expected behavior

Use looked up ip as parameter for external-check scripts (and port)

## Do you have any idea what may have caused this?

It's the combination of dnsresolvers and external-check

## Do you have an idea how to solve the issue?

Fix the code

**Comment by dfex55** [comment]
Tested down to v1.9 with same behaviour

**Comment by capflam** [comment]
Thanks for report. Indeed there is a bug. I pushed a fix today.

**Comment by dfex55** [comment]
Great news. Thanks!

little off:
When do you plan the next update to the docker hub (dev) images? And which tag will include the fix?

**Comment by capflam** [comment]
We are not responsible of the docker images. @TimWolla seems to contribute to it. I guess it will be part of the 2.2-dev7 (not released yet). The fix should still be backported in older branches. But for now there is no release planned.
In short, I don't know :)

**Comment by TimWolla** [comment]
> @TimWolla seems to contribute to it.

Yes, I usually do if I notice that something could be improved. Not only for HAProxy, but all across the board.

For HAProxy the following applies: The repository at [docker-library/haproxy](https://github.com/docker-library/haproxy) is automatically updated for any new HAProxy releases within existing branches (using `update.sh`). New version branches are added manually (usually by myself). Afterwards a PR is opened in [docker-library/official-images](https://github.com/docker-library/official-images/pulls) and the new image published once the PR is merged.

In short: The `haproxy` image on Docker Hub should follow HAProxy releases with a delay of one or two days.

**Comment by capflam** [comment]
No critical enough to be backported to 1.7 and 1.6. I close the issue keeping labels for the reference.
